# Исследование надежности заемщиков


## Откройте таблицу и изучите общую информацию о данных

**Задание 1. Импортируйте библиотеку pandas. Считайте данные из csv-файла в датафрейм и сохраните в переменную `data`. Путь к файлу:**

`/datasets/data.csv`

In [ ]:
import pandas as pd

try:
    data = pd.read_csv('/datasets/data.csv')
except:
    data = pd.read_csv('https://code.s3.yandex.net/datasets/data.csv')

**Задание 2. Выведите первые 20 строчек датафрейма `data` на экран.**

In [ ]:
data.head(20)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   
15         1    -972.364419         26              среднее             1   
16         0   -1719.934226         35              среднее             1   
17         0   -2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0  -10038.818549         48              СРЕДНЕЕ             1   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   
15   женат / замужем                 0      F   сотрудник     0   
16   женат / замужем                 0      F   сотрудник     0   
17  гражданский брак                 1      M   сотрудник     0   
18    вдовец / вдова                 2      F   пенсионер     0   
19         в разводе                 3      F   сотрудник     0   

     total_income                                 purpose  
0   253875.639453                           покупка жилья  
1   112080.014102                 приобретение автомобиля  
2   145885.952297                           покупка жилья  
3   267628.550329              дополнительное образование  
4   158616.077870                         сыграть свадьбу  
5   255763.565419                           покупка жилья  
6   240525.971920                       операции с жильем  
7   135823.934197                             образование  
8    95856.832424                   на проведение свадьбы  
9   144425.938277                 покупка жилья для семьи  
10  113943.491460                    покупка недвижимости  
11   77069.234271       покупка коммерческой недвижимости  
12            NaN                         сыграть свадьбу  
13  130458.228857                 приобретение автомобиля  
14  165127.911772              покупка жилой недвижимости  
15  116820.904450 

**Задание 3. Выведите основную информацию о датафрейме с помощью метода `info()`.**

In [ ]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


## Предобработка данных

### Удаление пропусков

**Задание 4. Выведите количество пропущенных значений для каждого столбца. Используйте комбинацию двух методов.**

In [ ]:
data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

**Задание 5. В двух столбцах есть пропущенные значения. Один из них — `days_employed`. Пропуски в этом столбце вы обработаете на следующем этапе. Другой столбец с пропущенными значениями — `total_income` — хранит данные о доходах. На сумму дохода сильнее всего влияет тип занятости, поэтому заполнить пропуски в этом столбце нужно медианным значением по каждому типу из столбца `income_type`. Например, у человека с типом занятости `сотрудник` пропуск в столбце `total_income` должен быть заполнен медианным доходом среди всех записей с тем же типом.**

In [ ]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['total_income'].isna()), 'total_income'] = \
    data.loc[(data['income_type'] == t), 'total_income'].median()

### Обработка аномальных значений

**Задание 6. В данных могут встречаться артефакты (аномалии) — значения, которые не отражают действительность и появились по какой-то ошибке. таким артефактом будет отрицательное количество дней трудового стажа в столбце `days_employed`. Для реальных данных это нормально. Обработайте значения в этом столбце: замените все отрицательные значения положительными с помощью метода `abs()`.**

In [ ]:
data['days_employed'] = data['days_employed'].abs()

**Задание 7. Для каждого типа занятости выведите медианное значение трудового стажа `days_employed` в днях.**

In [ ]:
data.groupby('income_type')['days_employed'].agg('median')

income_type
безработный        366413.652744
в декрете            3296.759962
госслужащий          2689.368353
компаньон            1547.382223
пенсионер          365213.306266
предприниматель       520.848083
сотрудник            1574.202821
студент               578.751554
Name: days_employed, dtype: float64

У двух типов (безработные и пенсионеры) получатся аномально большие значения. Исправить такие значения сложно, поэтому оставьте их как есть. Тем более этот столбец не понадобится вам для исследования.

**Задание 8. Выведите перечень уникальных значений столбца `children`.**

In [ ]:
data['children'].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5], dtype=int64)

**Задание 9. В столбце `children` есть два аномальных значения. Удалите строки, в которых встречаются такие аномальные значения из датафрейма `data`.**

In [ ]:
data = data[(data['children'] != -1) & (data['children'] != 20)]

**Задание 10. Ещё раз выведите перечень уникальных значений столбца `children`, чтобы убедиться, что артефакты удалены.**

In [ ]:
data['children'].unique()

array([1, 0, 3, 2, 4, 5], dtype=int64)

### Удаление пропусков (продолжение)

**Задание 11. Заполните пропуски в столбце `days_employed` медианными значениями по каждого типа занятости `income_type`.**

In [ ]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['days_employed'].isna()), 'days_employed'] = \
    data.loc[(data['income_type'] == t), 'days_employed'].median()

**Задание 12. Убедитесь, что все пропуски заполнены. Проверьте себя и ещё раз выведите количество пропущенных значений для каждого столбца с помощью двух методов.**

In [ ]:
data.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

### Изменение типов данных

**Задание 13. Замените вещественный тип данных в столбце `total_income` на целочисленный с помощью метода `astype()`.**

In [ ]:
data['total_income'] = data['total_income'].astype(int)
data['children'] = data['children'].astype(int)
data['debt'] = data['debt'].astype(int)

### Обработка дубликатов

**Задание 14. Обработайте неявные дубликаты в столбце `education`. В этом столбце есть одни и те же значения, но записанные по-разному: с использованием заглавных и строчных букв. Приведите их к нижнему регистру. Проверьте остальные столбцы.**

In [ ]:
data['education'] = data['education'].str.lower()

**Задание 15. Выведите на экран количество строк-дубликатов в данных. Если такие строки присутствуют, удалите их.**

In [ ]:
data.duplicated().sum()

71

In [ ]:
data = data.drop_duplicates()

### Категоризация данных

**Задание 16. На основании диапазонов, указанных ниже, создайте в датафрейме `data` столбец `total_income_category` с категориями:**

- 0–30000 — `'E'`;
- 30001–50000 — `'D'`;
- 50001–200000 — `'C'`;
- 200001–1000000 — `'B'`;
- 1000001 и выше — `'A'`.


**Например, кредитополучателю с доходом 25000 нужно назначить категорию `'E'`, а клиенту, получающему 235000, — `'B'`. Используйте собственную функцию с именем `categorize_income()` и метод `apply()`.**

In [ ]:
def categorize_income(income):
    try:
        if 0 <= income <= 30000:
            return 'E'
        elif 30001 <= income <= 50000:
            return 'D'
        elif 50001 <= income <= 200000:
            return 'C'
        elif 200001 <= income <= 1000000:
            return 'B'
        elif income >= 1000001:
            return 'A'
    except:
        pass

In [ ]:
data['total_income_category'] = data['total_income'].apply(categorize_income)

**Задание 17. Выведите на экран перечень уникальных целей взятия кредита из столбца `purpose`.**

In [ ]:
data['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

**Задание 18. Создайте функцию, которая на основании данных из столбца `purpose` сформирует новый столбец `purpose_category`, в который войдут следующие категории:**

- `'операции с автомобилем'`,
- `'операции с недвижимостью'`,
- `'проведение свадьбы'`,
- `'получение образования'`.

**Например, если в столбце `purpose` находится подстрока `'на покупку автомобиля'`, то в столбце `purpose_category` должна появиться строка `'операции с автомобилем'`.**

**Используйте собственную функцию с именем `categorize_purpose()` и метод `apply()`. Изучите данные в столбце `purpose` и определите, какие подстроки помогут вам правильно определить категорию.**

In [ ]:
def categorize_purpose(row):
    try:
        if 'автом' in row:
            return 'операции с автомобилем'
        elif 'жил' in row or 'недвиж' in row:
            return 'операции с недвижимостью'
        elif 'свад' in row:
            return 'проведение свадьбы'
        elif 'образов' in row:
            return 'получение образования'
    except:
        return 'нет категории'

In [ ]:
data['purpose_category'] = data['purpose'].apply(categorize_purpose)

In [ ]:
data.info()

<class 'pandas.core.frame.DataFrame'>
Index: 21331 entries, 0 to 21524
Data columns (total 14 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   children               21331 non-null  int32  
 1   days_employed          21331 non-null  float64
 2   dob_years              21331 non-null  int64  
 3   education              21331 non-null  object 
 4   education_id           21331 non-null  int64  
 5   family_status          21331 non-null  object 
 6   family_status_id       21331 non-null  int64  
 7   gender                 21331 non-null  object 
 8   income_type            21331 non-null  object 
 9   debt                   21331 non-null  int32  
 10  total_income           21331 non-null  int32  
 11  purpose                21331 non-null  object 
 12  total_income_category  21331 non-null  object 
 13  purpose_category       21331 non-null  object 
dtypes: float64(1), int32(3), int64(3), object(7)
memory usage: 

### Шаг 3. Исследуйте данные и ответьте на вопросы

#### 3.1 Есть ли зависимость между количеством детей и возвратом кредита в срок?

Построим сводную таблицу `children_pivot` и добавим еще один столбец - `Доля должников`. В нем будет отражено отношение числа должников к общему числу заемщиков

In [ ]:
children_pivot = data.pivot_table('debt', index='children', aggfunc=['count', 'sum', 'mean'], fill_value=0)

In [ ]:
children_pivot = children_pivot.rename(columns={'count': 'Всего кредитополучателей', 'sum': 'Всего должников', 'mean': 'Доля должников', 'debt': ''})
children_pivot

Всего кредитополучателей Всего должников Доля должников
                                                                
children                                                        
0                           14091            1063       0.075438
1                            4808             444       0.092346
2                            2052             194       0.094542
3                             330              27       0.081818
4                              41               4       0.097561
5                               9               0       0.000000

Таблица состоит из столбцов `0` - количество человек без задолженностей, `1` - количество человек с задолженностью, `Доля должников` - отношение задолжников к общему числу заемщиков. Индексом мы взяли количество детей `children` (от 0 до 5)

**Вывод:** Из таблицы видно, что чаще вовремя платят кредит люди, не имеющие детей, реже - те, у кого 2 или 4 ребенка (стоит отметить, что для "5" детей данных недостаточно). **Зависимость есть**.

#### 3.2 Есть ли зависимость между семейным положением и возвратом кредита в срок?

Создадим сводную таблицу `family_status_pivot`, в которую добавим столбец `Доля должников`. В нем будет отражено отношение должников к общему числу заемщиков для каждой категории семейного положения.

In [ ]:
family_status_pivot = data.pivot_table(values='debt', index='family_status', aggfunc=['count', 'sum', 'mean'], fill_value=0)

In [ ]:
family_status_pivot = family_status_pivot.rename(columns={'count': 'Всего кредитополучателей', 'sum': 'Всего должников', 'mean': 'Доля должников', 'debt': ''})
family_status_pivot.sort_values(by='Доля должников', ascending=False)

Всего кредитополучателей Всего должников Доля должников
                                                                             
family_status                                                                
Не женат / не замужем                     2796             273       0.097639
гражданский брак                          4134             385       0.093130
женат / замужем                          12261             927       0.075606
в разводе                                 1189              84       0.070648
вдовец / вдова                             951              63       0.066246

**Вывод:** Из таблицы видим, что лучшие показатели платежспособности здесь у категории "вдовец/вдова" и "в разводе", которые относятся к срокам более ответственно, чем категории "Не женат/не замужем" и "гражданский брак" - у них худшие показатели. **Зависимость наблюдается**.

#### 3.3 Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

Создадим сводную таблицу `total_income_pivot`, в которую добавим столбец `Доля должников`. В нем будет отражено отношение должников к общему числу заемщиков для каждой категории доходов.

In [ ]:
total_income_pivot = data.pivot_table('debt', index='total_income_category', aggfunc=['count', 'sum', 'mean'], fill_value=0)

In [ ]:
total_income_pivot = total_income_pivot.rename(columns={'count': 'Всего кредитополучателей', 'sum': 'Всего должников', 'mean': 'Доля должников', 'debt': ''})
total_income_pivot.sort_values(by='Доля должников', ascending=False)

Всего кредитополучателей Всего должников Доля должников
                                                                             
total_income_category                                                        
E                                           22               2       0.090909
C                                        15921            1353       0.084982
A                                           25               2       0.080000
B                                         5014             354       0.070602
D                                          349              21       0.060172

**Вывод:** Более платежеспособна по отношению к кредиту категория D (диапазон дохода 30000 - 50000), менее - E (диапазон дохода 0 - 30000). Стоит отметить очень большую разницу в количестве человек между категориями. **Зависимость подтверждается**.

#### 3.4 Как разные цели кредита влияют на его возврат в срок?

Создадим сводную таблицу `purpose_category_pivot`, в которую добавим столбец `Доля должников`. В нем будет отражено отношение должников к общему числу заемщиков для каждой категории целей.

In [ ]:
purpose_category_pivot = data.pivot_table(values='debt', index='purpose_category', aggfunc=['count', 'sum', 'mean'], fill_value=0)

In [ ]:
purpose_category_pivot = purpose_category_pivot.rename(columns={'count': 'Всего кредитополучателей', 'sum': 'Всего должников', 'mean': 'Доля должников', 'debt': ''})
purpose_category_pivot.sort_values(by='Доля должников', ascending=False)

Всего кредитополучателей Всего должников  \
                                                                    
purpose_category                                                    
операции с автомобилем                       4279             400   
получение образования                        3988             369   
проведение свадьбы                           2313             183   
операции с недвижимостью                    10751             780   

                         Доля должников  
                                         
purpose_category                         
операции с автомобилем         0.093480  
получение образования          0.092528  
проведение свадьбы             0.079118  
операции с недвижимостью       0.072551

**Вывод:** Люди из категории "операции с недвижимостью" более ответственно подходят к оплате кредита в срок, нежели люди из категории "операции с автомобилем", хотя и разница между конверсиями небольшая.  

#### 3.5 Приведите возможные причины появления пропусков в исходных данных.

*Ответ:* Есть 3 вида пропусков:
* Полностью случайные. Верояность встретить такие пропуски не зависит ни от каких других значений. Такие пропуски можно восстановить по имени
* Случайные. Вероятность пропуска зависит от других значений в наборе данных, но не от собственного столбца. Например, пропущенное значение связано с тем, что такой категории не существует
* Неслучайные. Вероятность пропуска зависит от других значений, в том числе и от значений собственного столбца. Отсутствующее значение зависит от характера вопроса - такие значения пропускают специально - и от значения в другом столбце

#### 3.6 Объясните, почему заполнить пропуски медианным значением — лучшее решение для количественных переменных.

*Ответ:* На медианное значение не влияют выбивающиеся из общей массы числа, в отличие от среднего значения. Например, при расчете дохода лучше использовать медиану, поскольку если между доходами существенный разрыв, то среднее значение сместится в сторону большего числа

### Шаг 4: общий вывод.

Датасет состоит из 12 столбцов и из 21525 строк. Типы данных - int64, float64, object.
В ходе исследования надежности заемщиков были выполнены следующие шаги:
* Обработны пропуски в двух столбцах: `days_employed` и `total_income`. Для заполнения пропусков мы взяли медиану для каждой категории сотрудников. Также, в `days_employed` отрицательные значения были заменены на положительные;
* Устранены аномальные значения в столбце `children` (-1 и 20);
* Вещественный тип данных в столбце `total_income` для удобства был заменен на целочисленный;
* Обработаны неявные дубликаты в столбце `education` - c помощью приведения к нижнему регистру были найдены дубликаты и удалены;
* Категоризированы данные для целей взятия кредита и для уровней дохода с помощью собственных функций. Добавлены соответствующие столбцы `purpose_category` и `total_income_category`
___
Для удобства анализа были построены сводные таблицы `children_pivot`, `family_status_pivot`, `total_income_pivot` и `purpose_category_pivot`. Исходя из таблиц проверены следующие гипотезы:
* Есть ли зависимость между количеством детей и возвратом кредита в срок? **Да, есть**. Как показал анализ, лучшие показатели у клиентов без детей (7.54 %), худшие - люди из категории "4" (9.76 %)
* Есть ли зависимость между семейным положением и возвратом кредита в срок? **Да, есть**. Лучшие показатели платежспособности у категории "вдовец/вдова" (6.62 %) и "в разводе" (7.06 %), которые относятся к срокам более ответственно, чем категории "Не женат/не замужем" (9.76 %) и "гражданский брак" (9.31 %) - у них худшие показатели
* Есть ли зависимость между уровнем дохода и возвратом кредита в срок? **Да, есть**. Более платежеспособна по отношению к кредиту категория D (диапазон дохода 30000 - 50000) - 6.02 %, менее - E (диапазон дохода 0 - 30000) - 9.09 %.

Что касается вопроса "Как разные цели кредита влияют на его возврат в срок?" - самый большой процент должников оказался у людей из категории "операции с автомобилем" - `9.35%`, самый маленький процент у категории "операции с недвижимостью" - `7.26%`.
___
Портрет добросовестного заемщика:
**Вдовец/вдова без детей, взявший/-ая кредит на недвижимость, и с доходом от 30000 до 50000.**